### PR TITLE
Prevent memory leak from reoccurring when presence is disabled.

### PR DESCRIPTION
**Pull request by erikjohnston**
When presence was disabled in Synapse <v1.58.0 (fixed by #12213) there was a small memory leak where the sync code would call `set_state` and insert items into the wheel timer, but would then never read *from* the wheel timer, causing a memory leak.

### CODE REVIEW

File: synapse/handlers/presence.py [hunk 1]
```diff
@@ -659,27 +659,28 @@ def __init__(self, hs: "HomeServer"):
...
-            self.wheel_timer.insert(
-                now=now, obj=state.user_id, then=state.last_active_ts + IDLE_TIMER
-            )
-            self.wheel_timer.insert(
-                now=now,
-                obj=state.user_id,
-                then=state.last_user_sync_ts + SYNC_ONLINE_TIMEOUT,
-            )
-            if self.is_mine_id(state.user_id):
+        if self._presence_enabled:
```
Comment (DMRobertson) [inline synapse/handlers/presence.py:662]: Feels a bit weird, Would it maybe make more sense to not register the prescence handler if presence is disabled? :shrug:

File: synapse/util/wheel_timer.py [hunk 2]
```diff
@@ -84,7 +87,7 @@ def fetch(self, now: int) -> List[T]:
         """
         now_key = int(now / self.bucket_size)
 
-        ret = []
+        ret: List[T] = []
         while self.entries and self.entries[0].end_key <= now_key:
             ret.extend(self.entries.pop(0).queue)
```
Comment (DMRobertson) [inline synapse/util/wheel_timer.py:105]: I was going to say "where do we remove elements from `_Entry.queue`. Looks like we consume each queue/bucket as a whole and them drop them for the GC to reclaim.

File: synapse/util/wheel_timer.py [hunk 3]
```diff
@@ -11,17 +11,20 @@
...
 class _Entry(Generic[T]):
     __slots__ = ["end_key", "queue"]
 
     def __init__(self, end_key: int) -> None:
         self.end_key: int = end_key
-        self.queue: List[T] = []
+
+        # We use a set here as otherwise we can end up with a lot of duplicate
+        # entries.
+        self.queue: Set[T] = set()
```
Comment (DMRobertson) [inline synapse/util/wheel_timer.py:30]: maybe `self.elements` instead of `self.queue`, since this is no longer ordered?

**Review by DMRobertson — APPROVED** [review]
A couple of minor things I'm not hugely thrilled about, but I'll leave those to your discretion. LGTM as is.

File: synapse/handlers/presence.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/handlers/presence.py:662]: Yeah, though note we do `self._presence_enabled` checks in a lot of places already. We could change the way we handle presence being disabled (and I agree that it'd quite probably be nicer to just never have a `PresenceHandler`), but that is a refactor I'd rather not take on here.

File: synapse/util/wheel_timer.py (hunk 3 above)
Comment (erikjohnston) [reply synapse/util/wheel_timer.py]: Good idea, done!

File: synapse/handlers/presence.py (hunk 1 above)
Comment (DMRobertson) [reply synapse/handlers/presence.py:662]: extremely reasonable, perfectly proper, utterly understandable, I can't think of more synonyms for "very fair"

File: synapse/util/wheel_timer.py [hunk 4]
```diff
@@ -14,20 +14,17 @@
...
 class _Entry(Generic[T]):
-    __slots__ = ["end_key", "queue"]
-
-    def __init__(self, end_key: int) -> None:
-        self.end_key: int = end_key
-
-        # We use a set here as otherwise we can end up with a lot of duplicate
-        # entries.
-        self.queue: Set[T] = set()
+    end_key: int
```
Comment (DMRobertson) [inline synapse/util/wheel_timer.py:26]: What's `end_key`? Oh, it's a timestamp?

Comment (DMRobertson) [reply synapse/util/wheel_timer.py:26]: Oh, but they're not necessarily timestamps, just arbitrary integers for counting _something_.

**Review by DMRobertson — APPROVED** [review]
lgtm. :ship: :it:

**Pull request merged**